external/fv3fit/fv3fit/keras/_models/convolutional.py

```diff
@@ -0,0 +1,194 @@
+
+    Returns a dict containing "dims" and "units" for each name.
+    """
+    metadata = []
+    for name in names:
+        metadata.append(
+            {"dims": ds[name].dims, "units": ds[name].attrs.get("units", "unknown")}
+        )
+    return tuple(metadata)
+
+
+@dataclasses.dataclass
+class ConvolutionalHyperparameters(Hyperparameters):
+    """
+    Args:
+        input_variables: names of variables to use as inputs
+        output_variables: names of variables to use as outputs
+        optimizer_config: selection of algorithm to be used in gradient descent
+        convolutional_network: configuration of convolutional network
+        training_loop: configuration of training loop
+        loss: configuration of loss functions, will be applied separately to
+            each output variable
+    """
+
+    input_variables: List[str]
+    output_variables: List[str]
+    optimizer_config: OptimizerConfig = dataclasses.field(
+        default_factory=lambda: OptimizerConfig("Adam")
+    )
+    convolutional_network: ConvolutionalNetworkConfig = dataclasses.field(
+        default_factory=lambda: ConvolutionalNetworkConfig()
+    )
+    training_loop: TrainingLoopConfig = dataclasses.field(
+        default_factory=lambda: TrainingLoopConfig(epochs=10)
+    )
+    loss: LossConfig = LossConfig(scaling="standard")
+
+    @property
+    def variables(self) -> Set[str]:
+        return set(self.input_variables).union(self.output_variables)
+
+
+@register_training_function("convolutional", ConvolutionalHyperparameters)
+def train_convolutional_model(
+    hyperparameters: ConvolutionalHyperparameters,
+    train_batches: Sequence[xr.Dataset],
+    validation_batches: Sequence[xr.Dataset],
+):
+    validation_data = batch_to_array_tuple(
+        validation_batches[0],
+        input_variables=hyperparameters.input_variables,
+        output_variables=hyperparameters.output_variables,
+    )
+    train_data = _XyMultiArraySequence(
+        X_names=hyperparameters.input_variables,
+        y_names=hyperparameters.output_variables,
+        dataset_sequence=train_batches,
+    )
+    train_model, predict_model = build_model(hyperparameters, batch=train_batches[0])
+    hyperparameters.training_loop.fit_loop(
+        model=train_model, Xy=train_data, validation_data=validation_data
+    )
+    output_metadata = get_metadata(
+        names=hyperparameters.output_variables, ds=validation_batches[0]
+    )
+    predictor = PureKerasNoStackModel(
+        input_variables=hyperparameters.input_variables,
+        output_variables=hyperparameters.output_variables,
+        output_metadata=output_metadata,
+        model=predict_model,
+    )
+    return predictor
+
+
+def batch_to_array_tuple(
+    batch: xr.Dataset, input_variables: Sequence[str], output_variables: Sequence[str]
+) -> Tuple[Tuple[np.ndarray, ...], Tuple[np.ndarray, ...]]:
+    return (
+        tuple(batch[name].values for name in input_variables),
```

Comment:
Maybe I missed this somewhere else, is there an assumption here about the dimension ordering in the dataset such that the last dimension is z? Is the dimension order guaranteed to be (..., z)?

Reply:
I think the no-stacking limitation is fine as long as the assumptions are clearly documented. I was mainly wondering if it was possible for a dataset to be provided as input with some other ordering than x, y, z and have this result in some silent bug in the results or fail with a confusing error.

Reply:
I had to add the stacking anyways to get the tests to pass after changing some other things, so this is no longer an issue. it still assumes the non-stacked dimension ordering.